programs/library-bridge/CatBoostLibraryHandler.cpp

```diff
@@ -0,0 +1,376 @@
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int BAD_ARGUMENTS;
+    extern const int CANNOT_APPLY_CATBOOST_MODEL;
+    extern const int CANNOT_LOAD_CATBOOST_MODEL;
+    extern const int LOGICAL_ERROR;
+}
+
+CatBoostLibraryHandler::APIHolder::APIHolder(SharedLibrary & lib)
+{
+    ModelCalcerCreate = lib.get<CatBoostLibraryAPI::ModelCalcerCreateFunc>(CatBoostLibraryAPI::ModelCalcerCreateName);
+    ModelCalcerDelete = lib.get<CatBoostLibraryAPI::ModelCalcerDeleteFunc>(CatBoostLibraryAPI::ModelCalcerDeleteName);
+    GetErrorString = lib.get<CatBoostLibraryAPI::GetErrorStringFunc>(CatBoostLibraryAPI::GetErrorStringName);
+    LoadFullModelFromFile = lib.get<CatBoostLibraryAPI::LoadFullModelFromFileFunc>(CatBoostLibraryAPI::LoadFullModelFromFileName);
+    CalcModelPredictionFlat = lib.get<CatBoostLibraryAPI::CalcModelPredictionFlatFunc>(CatBoostLibraryAPI::CalcModelPredictionFlatName);
+    CalcModelPrediction = lib.get<CatBoostLibraryAPI::CalcModelPredictionFunc>(CatBoostLibraryAPI::CalcModelPredictionName);
+    CalcModelPredictionWithHashedCatFeatures = lib.get<CatBoostLibraryAPI::CalcModelPredictionWithHashedCatFeaturesFunc>(CatBoostLibraryAPI::CalcModelPredictionWithHashedCatFeaturesName);
+    GetStringCatFeatureHash = lib.get<CatBoostLibraryAPI::GetStringCatFeatureHashFunc>(CatBoostLibraryAPI::GetStringCatFeatureHashName);
+    GetIntegerCatFeatureHash = lib.get<CatBoostLibraryAPI::GetIntegerCatFeatureHashFunc>(CatBoostLibraryAPI::GetIntegerCatFeatureHashName);
+    GetFloatFeaturesCount = lib.get<CatBoostLibraryAPI::GetFloatFeaturesCountFunc>(CatBoostLibraryAPI::GetFloatFeaturesCountName);
+    GetCatFeaturesCount = lib.get<CatBoostLibraryAPI::GetCatFeaturesCountFunc>(CatBoostLibraryAPI::GetCatFeaturesCountName);
+    GetTreeCount = lib.tryGet<CatBoostLibraryAPI::GetTreeCountFunc>(CatBoostLibraryAPI::GetTreeCountName);
+    GetDimensionsCount = lib.tryGet<CatBoostLibraryAPI::GetDimensionsCountFunc>(CatBoostLibraryAPI::GetDimensionsCountName);
+}
+
+CatBoostLibraryHandler::CatBoostLibraryHandler(
+    const std::string & library_path,
+    const std::string & model_path)
+    : library(std::make_shared<SharedLibrary>(library_path))
+    , api(*library)
+{
+    model_calcer_handle = api.ModelCalcerCreate();
+
+    if (!api.LoadFullModelFromFile(model_calcer_handle, model_path.c_str()))
+    {
+        throw Exception(ErrorCodes::CANNOT_LOAD_CATBOOST_MODEL,
+                "Cannot load CatBoost model: {}", api.GetErrorString());
+    }
+
+    float_features_count = api.GetFloatFeaturesCount(model_calcer_handle);
+    cat_features_count = api.GetCatFeaturesCount(model_calcer_handle);
+
+    tree_count = 1;
+    if (api.GetDimensionsCount)
+        tree_count = api.GetDimensionsCount(model_calcer_handle);
+}
+
+CatBoostLibraryHandler::~CatBoostLibraryHandler()
+{
+    api.ModelCalcerDelete(model_calcer_handle);
+}
+
+namespace
+{
+
+/// Buffer should be allocated with features_count * column->size() elements.
```

Comment:
The code from here until the end of the file is (almost) 1:1 copypasted from https://clickhouse.com/codebrowser/ClickHouse/src/Interpreters/CatBoostModel.cpp.html#_ZNK2DB17CatBoostModelImpl8evaluateERKNSt3__16vectorIPKNS_7IColumnENS1_9allocatorIS5_EEEE